# AEM Basic verification 1 Beam

In an attempt to validate the implementation of the AEM in a python program called AEM_basic.py the following input decks were written and then compared to the same problem modeled using the FEM with Calculix as well as the analytical solution expected for the problem.

## Uniaxial Loading of beam

Uniaxial load is applied to a beam end of 10kN.  The displacement is compared to the same problem modeled using FEM and to an analytical solution.

### Analytical Solution

Since there is no change in area the stress remains constant over the beam $$ \sigma_{xx} = \frac{F}{A} $$ The stress can then be related to the strain using Hooke's law: $$ \sigma = E \epsilon $$ with E = 207 GPa Now relating the strain to the displacement of the tip of the beam where the distributed force is applied $$ \epsilon_{xx} = \frac{\partial u_{x}}{\partial x} = \frac{\frac{F}{A}}{E}$$ $$ \frac{dU_x}{dx} = \frac{F}{AE} $$ giving $$ \int_{0}^{x} du_x = \int_{0}^{x} \frac{F}{AE} dx $$ After integration and using boundary conditions knowing the displacement of the beam at x = 0 the displacement at the bar end is given as $$ \delta_{end} = \frac{FL}{EA} $$ or at ant other point $$ \delta_x = \frac{Fx}{EA} $$
For the given problem $$ \delta_{end} = \frac{100 000 \times 2}{207 \times 10^{9} \times 0.2 \times 0.15} $$
$$ \delta_{end} = 3.2206 \times 10^{-5}$$

### AEM Solution using 2x1

The grid used is 2 elements in the x direction and one in the y.

In [1]:
import numpy as np
import sys
sys.path.insert(0, '../..')
from AEM_basic import AEM_basic

## Start of input deck
# Material properties
E = 207e+9
nu = 0.3
G = E/(2*(1+nu))
T = 0.15

# Element properties
#& Should be adjusted to load from hex_mesher
grid = [2, 1, 0]
num_ele = grid[0]*grid[1]
el_center = np.zeros(shape=(num_ele, 3))
el_count = 0
for y_c in range(100, 200, 100):
    for x_c in range(500, 2000, 1000):
        el_center[el_count, 0] = el_count
        el_center[el_count, 1] = float(x_c)/1000
        el_center[el_count, 2] = float(y_c)/1000
        el_count += 1


a1 = 1.        # width of element 1
b1 = 0.200         # height of element 1
theta1 = 0.      # radians (for time being not applicable)

#@ Seems redundent unless different size and shape elements
#& Remove in future releases
a2 = 1.         # width of element 2
b2 = 0.200         # height of element 2
theta2 = 0.      # radians (for time being not applicable)

gap = [0.00, 0.00, 0]         # size of gap between elements

# Spring properties
#@ Assuming same number of connecting springs for horizontal and vetical
num_spring = 10

# Prescribed displacements
    # element number, DOF, value
Up = np.array([0, 0, 0])
#@ Needs to be done to read from input
dof = np.ones(num_ele*3)
dof[0] = 0
dof[1] = 0
dof[2] = 0

# Prescribed forces
    # element number, DOF, value
F = np.zeros(shape=(num_ele*3, 1))
F[3, 0] = 1.0e+5

## End of input deck

U, F, strain_ele, stress_Pstrain, stress_Pstress = AEM_basic(E, nu, G, T, grid, num_ele,
              el_center, a1, a2, b1, b2, theta1, theta2, gap,
              num_spring, Up, dof, F)

print('U=')
#print(U[-3, 0])
print(U)
#print('F=')
#print(F)
#print('strain=')
#print(strain_ele)
#print('Pstrain=')
#print(stress_Pstrain)
#print('Pstress=')
#print(stress_Pstress)

U=
[[  0.00000000e+00]
 [  0.00000000e+00]
 [  0.00000000e+00]
 [  1.61030596e-05]
 [ -2.92727134e-21]
 [ -5.85454269e-21]]


### AEM Solution 10x1

In [6]:
import numpy as np
import sys
sys.path.insert(0, '../..')
from AEM_basic import AEM_basic

## Start of input deck
# Material properties
E = 207e+9
nu = 0.3
G = E/(2*(1+nu))
T = 0.15

# Element properties
#& Should be adjusted to load from hex_mesher
grid = [10, 1, 0]
num_ele = grid[0]*grid[1]
el_center = np.zeros(shape=(num_ele, 3))
el_count = 0
for y_c in range(100, 200, 100):
    for x_c in range(100, 2000, 200):
        el_center[el_count, 0] = el_count
        el_center[el_count, 1] = float(x_c)/1000
        el_center[el_count, 2] = float(y_c)/1000
        el_count += 1


a1 = 0.2         # width of element 1
b1 = 0.200         # height of element 1
theta1 = 0.      # radians (for time being not applicable)

#@ Seems redundent unless different size and shape elements
#& Remove in future releases
a2 = 0.2         # width of element 2
b2 = 0.200         # height of element 2
theta2 = 0.      # radians (for time being not applicable)

gap = [0.00, 0.00, 0]         # size of gap between elements

# Spring properties
#@ Assuming same number of connecting springs for horizontal and vetical
num_spring = 10

# Prescribed displacements
    # element number, DOF, value
Up = np.array([0, 0, 0])
#@ Needs to be done to read from input
dof = np.ones(num_ele*3)
dof[0] = 0
dof[1] = 0
dof[2] = 0

# Prescribed forces
    # element number, DOF, value
F = np.zeros(shape=(num_ele*3, 1))
F[27, 0] = 1.0e+5

## End of input deck

U, F, strain_ele, stress_Pstrain, stress_Pstress = AEM_basic(E, nu, G, T, grid, num_ele,
              el_center, a1, a2, b1, b2, theta1, theta2, gap,
              num_spring, Up, dof, F)

print('U=')
print(U[-3, 0])
#print('F=')
#print(F)
#print('strain=')
#print(strain_ele)
#print('Pstrain=')
#print(stress_Pstrain)
#print('Pstress=')
#print(stress_Pstress)

U=
2.89855072464e-05


#### Error in U

Due to the nature of the AEM applied boundary conditions the length depicted aren't the actual length of the beam.  The "first" half of element 0 and the "last" half of element 9 aren't included in the calculations.  Resulting in an expected error in displacement of 10% $$ \frac{0.1+0.1}{2} \times 100 $$

Actual error:  $$ \frac{3.2206 \times 10^{-05} - 2.8986 \times 10^{-05}}{3.2206 \times 10^{-05}} \times 100 = 9.998\% $$

### AEM Solution 2x2

In [1]:
import numpy as np
import sys
sys.path.insert(0, '../..')
from AEM_basic import AEM_basic

## Start of input deck
# Material properties
E = 207e+9
nu = 0.3
G = E/(2*(1+nu))
T = 0.15

# Element properties
#& Should be adjusted to load from hex_mesher
grid = [2, 2, 0]
num_ele = grid[0]*grid[1]
el_center = np.zeros(shape=(num_ele, 3))
el_count = 0
for y_c in range(50, 200, 100):
    for x_c in range(500, 2000, 1000):
        el_center[el_count, 0] = el_count
        el_center[el_count, 1] = float(x_c)/1000
        el_center[el_count, 2] = float(y_c)/1000
        el_count += 1


a1 = 1.         # width of element 1
b1 = 0.100         # height of element 1
theta1 = 0.      # radians (for time being not applicable)

#@ Seems redundent unless different size and shape elements
#& Remove in future releases
a2 = 1.         # width of element 2
b2 = 0.100         # height of element 2
theta2 = 0.      # radians (for time being not applicable)

gap = [0.00, 0.00, 0]         # size of gap between elements

# Spring properties
#@ Assuming same number of connecting springs for horizontal and vetical
num_spring = 10

# Prescribed displacements
    # element number, DOF, value
Up = np.array([0, 0, 0, 0, 0, 0])
#@ Needs to be done to read from input
dof = np.ones(num_ele*3)
dof[0] = 0
dof[1] = 0
dof[2] = 0
dof[6] = 0
dof[7] = 0
dof[8] = 0

# Prescribed forces
    # element number, DOF, value
F = np.zeros(shape=(num_ele*3, 1))
F[3, 0] = 0.50e+5
F[9, 0] = 0.50e+5

## End of input deck

U, F, strain_ele, stress_Pstrain, stress_Pstress = AEM_basic(E, nu, G, T, grid, num_ele,
              el_center, a1, a2, b1, b2, theta1, theta2, gap,
              num_spring, Up, dof, F)

print('U=')
print(U)
#print('F=')
#print(F)
#print('strain=')
#print(strain_ele)
#print('Pstrain=')
#print(stress_Pstrain)
#print('Pstress=')
#print(stress_Pstress)

U=
[[  0.00000000e+00]
 [  0.00000000e+00]
 [  0.00000000e+00]
 [  1.58915753e-05]
 [ -2.14414886e-06]
 [ -6.17419334e-06]
 [  0.00000000e+00]
 [  0.00000000e+00]
 [  0.00000000e+00]
 [  1.63145438e-05]
 [ -2.14052213e-06]
 [ -2.39514864e-06]]


#### Error in U

After the successful completion of the 10x1 beam implementation I tried a 10x2gird but the displacements at the beam tip did not react as expected.  I then tried a 2x2 solution and as can be seen in above results the displacements of the two elements at the beam tip is not the same in the x-direction.  Also in contrast to the original implementation of 2x1 the displacements in the y direction along with the rotation of the elements are allot higher than expected.  Obviously some mistake is made in the implementation of the vertical stiffness interactions.